### PR TITLE
Issue 4879: (KeyValueTables) List and Delete API for Key Value Tables on Controller

### LINKED ISSUE #4879: List and Delete APIs for KeyValueTable on Controller
**Problem description**
Provide API on Controller to list all KeyValueTables inside a scope and to delete a KeyValueTable.

**Problem location**
Controller

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by pbelgundi**
Signed-off-by: pbelgundi <prajakta.belgundi@emc.com>

Change log description
This PR has code to support List and Delete APIs on KeyValueTables in Pravega Controller.
For details please refer https://github.com/pravega/pravega/wiki/PDP-39-Key-Value-Tables

Purpose of the change
Fixes #4879

What the code does
Adds listKeyValueTablesInScope() API and deleteKeyValueTable() API to Controller, that enables client to list all KeyValueTables in the given scope, in a paginated fashion and API to delete a KeyValueTable.

How to verify it
All Unit and Integration Tests should pass.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java [hunk 1]
```diff
@@ -120,6 +120,6 @@ public static String ofReaderGroupInScope(String scopeName, String readerGroupNa
      */
     public static String ofKeyValueTableInScope(String scopeName, String kvtName) {
         Exceptions.checkNotNullOrEmpty(kvtName, "KeyValueTableName");
-        return String.format("%s/%s", ofStreamsInScope(scopeName), kvtName);
+        return String.format("%s/%s_kvt", ofStreamsInScope(scopeName), kvtName);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: you shouldnt add the suffix here. This is the authorization resource against which the auth rules will be set outside of pravega. 
the resource name being scope/tablename is sufficient.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java [hunk 2]
```diff
@@ -466,10 +466,10 @@ public String retrieveDelegationToken() {
             }
         });
     }
-
+    
     @Override
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java:470]: revert this class

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java [hunk 3]
```diff
@@ -144,6 +144,66 @@ public void getCurrentSegmentsKeyValueTable(KeyValueTableInfo request, StreamObs
...
+                        .handle((response, ex) -> {
+                            if (ex != null) {
+                                if (Exceptions.unwrap(ex) instanceof StoreException.DataNotFoundException) {
+                                    return Controller.KVTablesInScopeResponse.newBuilder().setStatus(Controller.KVTablesInScopeResponse.Status.SCOPE_NOT_FOUND).build();
+                                } else {
+                                    throw new CompletionException(ex);
+                                }
+                            } else {
+                                log.debug("All KeyValueTables in scope with continuation token: {}", response);
+                                List<KeyValueTableInfo> kvtables = response
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:183]: There is one problem with the logic here. And i realize this same problem is present for streams too. 
If we filter out unauthorized tables, then the response we send back may endup being empty while there may still be more records to be read from the continuation token.

So any such filtering should invoke the call with `filtered-out` elements and keep doing it until we have collected `limit` number of elements or have nothing to read anymore.

File: controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java (hunk 1 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: If there is a stream and kvtable with the same name inside the same scope, earlier code would generate the same "resource name" for both causing confusion.... Hence I think it would be good to have this suffix so we know which resource this is actually for.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java (hunk 2 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java:470]: This class is used by Integration Tests when invoking Controller operations.
Hence added listKeyValueTables() here too. 
Integration tests need to be fixed to use ControllerImpl.java in client instead.

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 4]
```diff
@@ -1411,5 +1447,11 @@ public void createKeyValueTable(KeyValueTableConfig kvtConfig, RPCAsyncCallback<
             clientStub.withDeadlineAfter(timeoutMillis, TimeUnit.MILLISECONDS)
                     .createKeyValueTable(kvtConfig, callback);
         }
+
+        public void listKeyValueTables(KVTablesInScopeRequest request,
```
[the remark this answers is not in the document]
Comment (pbelgundi) [reply client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java:1451]: All other similar methods in this class are public too and I guess it doesn't matter as the enclosing class is "private"...will remove modifier in next commit..

File: controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java (hunk 1 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: @ravisharda thoughts?

Comment (ravisharda) [reply controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: Are tables direct members of the scope? 

If yes, it might help to disambiguate what the resource type is, for authorization purposes. Say, both a stream and a table have the same name `abc`. So, the resource representation for both shall be `myscope/abc`. So, if the admin sets up an ACL for a user account such that it has access to `myscope/abc`, it'll authorize the user to both the stream and the table, which i sundesirable. 

I'd suggest putting the `_kvt` as a prefix tag, instead of a suffix. And also simplifying the tag name. Something like: `_kvtable_abc`. 

On a side note, this also reminds me that we need a more consistent way of structuring these resource names. But, that's for another day.

Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: Yes. Tables are direct members of scope. 
Just like a scope can have streams under it, the same scope can also have KeyValueTables under it and there could be a stream and kvtable with the same name under a given scope.

Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/AuthResourceRepresentation.java:123]: Based on an offline discussion, we have decided to use the auth resource string `%s/kvtable/%s` for KeyValueTables

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java (hunk 3 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java]: Fixed this.

File: controller/src/main/java/io/pravega/controller/store/Scope.java [hunk 5]
```diff
@@ -67,6 +67,18 @@
      */
     void refresh();
 
+    /**
+     * A paginated api on the scope to get requested number of kvtables from under the scope
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/Scope.java:71]: "kvtables" -> "Key-Value Tables"

Same below.

File: controller/src/test/java/io/pravega/controller/server/v1/ZKControllerServiceImplTest.java [hunk 6]
```diff
@@ -238,4 +238,10 @@ public void getCurrentSegmentsKeyValueTableTest() {
         //TODO: Place holder till we fix the test
         assertTrue(true);
     }
+
+    @Test
+    public void kvtablesInScopeTest() {
+        //TODO: Place holder till we fix the test
```
Comment (andreipaduroiu) [inline controller/src/test/java/io/pravega/controller/server/v1/ZKControllerServiceImplTest.java:244]: When are we fixing the test?

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:3]: Please fix the license header.

File: controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java [hunk 8]
```diff
@@ -58,7 +59,7 @@
     protected KVTableMetadataStore kvtStore;
     protected TableMetadataTasks kvtMetadataTasks;
     protected SegmentHelper segmentHelperMock;
-    protected final ScheduledExecutorService executor = Executors.newScheduledThreadPool(10);
+    protected final ScheduledExecutorService executor = Executors.newScheduledThreadPool(30);
```
Comment (andreipaduroiu) [inline controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java:62]: Is this change needed?

File: shared/controller-api/src/main/java/io/pravega/shared/controller/event/kvtable/DeleteTableEvent.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
```
Comment (andreipaduroiu) [inline shared/controller-api/src/main/java/io/pravega/shared/controller/event/kvtable/DeleteTableEvent.java:3]: license header

Comment (pbelgundi) [reply shared/controller-api/src/main/java/io/pravega/shared/controller/event/kvtable/DeleteTableEvent.java:3]: What is the issue here?

File: controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java (hunk 8 above)
Comment (pbelgundi) [reply controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java:62]: Had changed when testing, its not needed. Will revert.

File: controller/src/main/java/io/pravega/controller/store/Scope.java (hunk 5 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/Scope.java]: fixed.

File: controller/src/test/java/io/pravega/controller/server/v1/ZKControllerServiceImplTest.java (hunk 6 above)
Comment (pbelgundi) [reply controller/src/test/java/io/pravega/controller/server/v1/ZKControllerServiceImplTest.java:246]: As discussed, these are tests for Zookeeper based implementation of KVTables and I won't be fixing them in this PR.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import io.pravega.controller.task.KeyValueTable.TableMetadataTasks;
+import io.pravega.controller.util.RetryHelper;
+import io.pravega.shared.controller.event.kvtable.DeleteTableEvent;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Request handler for performing scale operations received from requeststream.
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:29]: "delete operation on table stream"

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    public CompletableFuture<Void> execute(final DeleteTableEvent request) {
+        String scope = request.getScope();
+        String kvt = request.getKvtName();
+        long requestId = request.getRequestId();
+        String kvTableId = request.getTableId().toString();
+
+        return RetryHelper.withRetriesAsync(() -> getKeyValueTable(scope, kvt)
+                .thenCompose(table -> table.getId()).thenCompose(id -> {
+            if (!id.equals(kvTableId)) {
+                return CompletableFuture.completedFuture(null);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:59]: please log that this workflow completed because of this condition.

File: controller/src/main/java/io/pravega/controller/store/kvtable/PravegaTablesKVTable.java [hunk 12]
```diff
@@ -191,11 +191,14 @@ private String getMetadataTableName(String id) {
 
     @Override
     CompletableFuture<Void> createKVTableMetadata() {
-        return getId().thenCompose(id -> {
-            String metadataTable = getMetadataTableName(id);
-            return CompletableFuture.allOf(storeHelper.createTable(metadataTable))
-                    .thenAccept(v -> log.debug("kvtable {}/{} metadata table {} created", getScopeName(), getName(), metadataTable));
-        });
+        return getId().thenCompose(id -> CompletableFuture.allOf(storeHelper.createTable(getMetadataTableName(id)))
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/kvtable/PravegaTablesKVTable.java:194]: there only seems to be one future.. why CompletableFuture.allOf?

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java (hunk 10 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:29]: > In general, looks good to me. I left a couple of comments.
> 
> Also, please resolve the merge conflicts as I've updated the base branch with the latest master.
>

File: controller/src/main/java/io/pravega/controller/store/kvtable/PravegaTablesKVTable.java (hunk 12 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/kvtable/PravegaTablesKVTable.java]: fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java (hunk 11 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:60]: fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java (hunk 10 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java]: fixed

File: controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java (hunk 8 above)
Comment (pbelgundi) [reply controller/src/test/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasksTest.java]: Fixed

File: client/src/test/java/io/pravega/client/control/impl/ControllerImplTest.java [hunk 13]
```diff
@@ -955,6 +957,71 @@ public void getCurrentSegmentsKeyValueTable(Controller.KeyValueTableInfo request
...
+            public void listKeyValueTablesInScope(Controller.KVTablesInScopeRequest request,
+                                                  StreamObserver<Controller.KVTablesInScopeResponse> responseObserver) {
+                if (request.getScope().getScope().equals(NON_EXISTENT)) {
+                    responseObserver.onNext(Controller.KVTablesInScopeResponse
+                            .newBuilder().setStatus(Controller.KVTablesInScopeResponse.Status.SCOPE_NOT_FOUND)
+                            .build());
+                    responseObserver.onCompleted();
+                } else if (request.getScope().getScope().equals("deadline")) {
+                    // dont send any response
+                    System.err.println("i am here");
```
Comment (RaulGracia) [inline client/src/test/java/io/pravega/client/control/impl/ControllerImplTest.java:971]: Please, remove think kind of console prints from the code.

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 14]
```diff
@@ -14,31 +14,7 @@
...
-import static io.pravega.shared.MetricsNames.TRUNCATE_STREAM_LATENCY;
-import static io.pravega.shared.MetricsNames.UPDATE_STREAM;
-import static io.pravega.shared.MetricsNames.UPDATE_STREAM_FAILED;
-import static io.pravega.shared.MetricsNames.UPDATE_STREAM_LATENCY;
-import static io.pravega.shared.MetricsNames.CREATE_KVTABLE_LATENCY;
-import static io.pravega.shared.MetricsNames.CREATE_KVTABLE;
-import static io.pravega.shared.MetricsNames.KVTABLE_SEGMENTS_COUNT;
-import static io.pravega.shared.MetricsNames.CREATE_KVTABLE_FAILED;
-import static io.pravega.shared.MetricsNames.globalMetricName;
+import static io.pravega.shared.MetricsNames.*;
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java:17]: Please, use single line imports as it was before.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    public CompletableFuture<Void> execute(final DeleteTableEvent request) {
+        String scope = request.getScope();
+        String kvt = request.getKvtName();
+        long requestId = request.getRequestId();
+        String kvTableId = request.getTableId().toString();
+
+        return RetryHelper.withRetriesAsync(() -> getKeyValueTable(scope, kvt)
+                .thenCompose(table -> table.getId()).thenCompose(id -> {
+            if (!id.equals(kvTableId)) {
+                log.debug("Skipped processing delete event for KeyValueTable {}/{} with Is:{} as UUIDs did not match.", scope, kvt, id);
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java:59]: `with Is` what does it mean?

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java [hunk 16]
```diff
@@ -144,6 +146,67 @@ public void getCurrentSegmentsKeyValueTable(KeyValueTableInfo request, StreamObs
...
     }
 
+    @Override
+    public void listKeyValueTablesInScope(Controller.KVTablesInScopeRequest request, StreamObserver<Controller.KVTablesInScopeResponse> responseObserver) {
+        String scopeName = request.getScope().getScope();
+        RequestTag requestTag = requestTracker.initializeAndTrackRequestTag(requestIdGenerator.get(),
+                "listKeyValueTables", scopeName);
+        log.info(requestTag.getRequestId(), "listKeyValueTables called for scope {}.", scopeName);
+
+        final AuthContext ctx;
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:156]: This can be simplified in a one-liner: `final AuthContext ctx = (this.grpcAuthHelper.isAuthEnabled()) ? AuthContext.current() : null;`

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java [hunk 17]
```diff
@@ -144,6 +146,67 @@ public void getCurrentSegmentsKeyValueTable(KeyValueTableInfo request, StreamObs
...
+            ctx = null;
+        }
+
+        authenticateExecuteAndProcessResults(
+                () -> {
+                    String result = this.grpcAuthHelper.checkAuthorization(
+                            AuthResourceRepresentation.ofScope(scopeName),
+                            AuthHandler.Permissions.READ,
+                            ctx);
+                    log.debug("Result of authorization for [{}] and READ permission is: [{}]",
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:169]: Do we have a consistent structure for the logging related to KV tables in controller? If not, it may be a good idea, so we can apply reggex expressions with `grep` when looking into the logs.

File: controller/src/main/java/io/pravega/controller/store/InMemoryScope.java [hunk 18]
```diff
@@ -147,4 +151,32 @@ public String getName() {
...
+        if (kvTablesMap == null) {
+            return Futures.failedFuture(StoreException.create(StoreException.Type.DATA_NOT_FOUND, this.scopeName));
+        }
+        List<String> sortedKVTablesList = kvTablesMap.keySet().stream().collect(Collectors.toList());
+        int start = 0, end = 0;
+        if (!continuationToken.isEmpty()) {
+            start = Integer.parseInt(continuationToken);
+        }
+
+        if ((start + limit) >= sortedKVTablesList.size()) {
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/store/InMemoryScope.java:166]: This can be simplified in a one-liner: `end = ((start + limit) >= sortedKVTablesList.size()) ? sortedKVTablesList.size() : start + limit;`

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java (hunk 14 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java:17]: This optimization is done by IntelliJ and is not done manually.

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java (hunk 17 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:169]: This method is used by both streams and KVTables.

File: client/src/test/java/io/pravega/client/control/impl/ControllerImplTest.java (hunk 13 above)
Comment (pbelgundi) [reply client/src/test/java/io/pravega/client/control/impl/ControllerImplTest.java]: removed

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java (hunk 14 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java (hunk 15 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/kvtable/DeleteTableTask.java]: typo fixed

File: controller/src/main/java/io/pravega/controller/store/InMemoryScope.java (hunk 18 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/InMemoryScope.java]: done

File: controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java (hunk 16 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java]: done
